source/test_env.cpp

```diff
@@ -14,115 +14,264 @@
-{
-    printf("{{%s;%s;%u}}" RCNL, TEST_ENV_MEASURE, measurement_name, value);
+void greentea_send_kv(const char *key) {
+    if (key) {
+        printf("{{%s;%d}}" NL, key, 0);
+    }
 }
 
-void notify_performance_coefficient(const char* measurement_name, const double value)
-{
-    printf("{{%s;%s;%f}}" RCNL, TEST_ENV_MEASURE, measurement_name, value);
+void notify_start() {
+    // Sync preamble: "{{__sync;0dad4a9d-59a3-4aec-810d-d5fb09d852c1}}"
+    // Example value of sync_uuid == "0dad4a9d-59a3-4aec-810d-d5fb09d852c1"
+	char _key[8] = {0};
+	char _value[48] = {0};
+	greentea_parse_kv(_key, _value, sizeof(_key), sizeof(_value));
+    greentea_send_kv(_key, _value);
 }
 
-void notify_completion(bool success)
-{
-    printf("{{%s}}" NL, success ? TEST_ENV_SUCCESS : TEST_ENV_FAILURE);
+void notify_timeout(const int timeout) {
+    greentea_send_kv(TEST_ENV_TIMEOUT, timeout);
+}
+
+void notify_hosttest(const char *host_test_name) {
+    greentea_send_kv(TEST_ENV_HOST_TEST_NAME, host_test_name);
+}
+
+void notify_completion(const int success) {
+    const char *val = success ? TEST_ENV_SUCCESS : TEST_ENV_FAILURE;
+    greentea_send_kv(TEST_ENV_END, val);
 #ifdef YOTTA_CFG_DEBUG_OPTIONS_COVERAGE
     coverage_report = true;
     gcov_exit();
     coverage_report = false;
 #endif
-    printf("{{%s}}" NL, TEST_ENV_END);
-    led_blink(LED1, success ? 1.0 : 0.1);
+    greentea_send_kv(TEST_ENV_EXIT, !success);
 }
 
-bool notify_completion_str(bool success, char* buffer)
-{
-    bool result = false;
-    if (buffer) {
-        sprintf(buffer, "{{%s}}" NL "{{%s}}" NL, success ? TEST_ENV_SUCCESS : TEST_ENV_FAILURE, TEST_ENV_END);
-        result = true;
-    }
-    return result;
+// Test Case support
+
+/** \brief Notifies test case start
+  * \param Test Case ID name
+  *
+  * This function notifies test environment abort test case execution start.
+  *
+  */
+void notify_testcase_start(const char *testcase_id) {
+    greentea_send_kv(TEST_ENV_TESTCASE_START, testcase_id);
 }
 
-// Host test auto-detection API
-void notify_host_test_name(const char *host_test) {
-    if (host_test) {
-        printf("{{host_test_name;%s}}" NL, host_test);
-    }
+/** \brief Return partial (test case) result from test suite
+  * \param Test Case ID name
+  * \param Success code, 0 - success, >0 failure reason, <0 inconclusive
+  *
+  * This function passes partial test suite's test case result to test
+  * environment.
+  * Each test suite (in many cases one binary with tests) can return
+  * multiple partial results used to track test case results.
+  *
+  * Test designers can use success code to return test case:
+  * success == 0 - PASS, test case execution was successful.
+  * success > 0  - FAILure, e.g. success == 404 can be used to
+  *                pass "Server not found".
+  * success < 0  - Inconclusive test case execution, e.g.
+  *
+  */
+void notify_testcase_finish(const char *testcase_id, const int success) {
```

Comment:
I think `result` instead of `success` is a better way of keeping my sanity intact.
I mean `success == 404` to signal a failure just looks really, really wrong ;-)